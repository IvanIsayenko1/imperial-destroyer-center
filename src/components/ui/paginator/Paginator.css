.paginator {
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 1rem;
  margin: 2rem 0;
  flex-wrap: wrap;
}

.paginator-button {
  padding: 0.5rem 1rem;
  background: rgba(255, 59, 59, 0.1);
  border: 1px solid rgba(255, 59, 59, 0.2);
  color: var(--rebel-base);
  border-radius: 4px;
  cursor: pointer;
  transition: all 0.2s ease;
  min-width: 100px;
}

.paginator-pages {
  display: flex;
  align-items: center;
  gap: 0.5rem;
  flex-wrap: wrap;
}

.paginator-page {
  min-width: 2.5rem;
  height: 2.5rem;
  border: none;
  background: transparent;
  color: var(--rebel-base);
  cursor: pointer;
  border-radius: 4px;
  transition: all 0.2s ease;
}

.paginator-page:hover:not(:disabled),
.paginator-button:hover:not(:disabled) {
  background: rgba(255, 59, 59, 0.2);
  box-shadow: 0 0 15px rgba(255, 59, 59, 0.3);
}

.paginator-page.active {
  background: rgba(255, 59, 59, 0.2);
  box-shadow: 0 0 10px rgba(255, 59, 59, 0.3);
}

.paginator-button:disabled,
.paginator-page:disabled {
  opacity: 0.5;
  cursor: not-allowed;
  box-shadow: none;
}

@media (max-width: 768px) {
  .paginator {
    flex-direction: column;
    gap: 1rem;
  }

  .paginator-pages {
    order: -1;
    justify-content: center;
  }

  .paginator-button {
    width: 100%;
    text-align: center;
  }

  .paginator-page {
    min-width: 2rem;
    height: 2rem;
  }
}

@media (max-width: 480px) {
  .paginator-pages {
    max-width: 100%;
    overflow-x: auto;
    padding: 0.5rem;
    scrollbar-width: none;
    -ms-overflow-style: none;
  }

  .paginator-pages::-webkit-scrollbar {
    display: none;
  }
}
