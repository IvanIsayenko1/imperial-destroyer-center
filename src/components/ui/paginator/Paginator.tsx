import React from 'react';
import './Paginator.css';

interface PaginatorProps {
  currentPage: number;
  totalElements: number;
  elementsPerPage: number;
  onPageChange: (page: number) => void;
}

export const Paginator: React.FC<PaginatorProps> = ({
  currentPage,
  totalElements,
  elementsPerPage,
  onPageChange,
}) => {
  const totalPages = Math.ceil(totalElements / elementsPerPage);
  const pageNumbers = Array.from({ length: totalPages }, (_, i) => i + 1);

  return (
    <div className="paginator box">
      <button
        className="paginator-button"
        onClick={() => onPageChange(currentPage - 1)}
        disabled={currentPage === 1}
      >
        Previous
      </button>

      <div className="paginator-pages">
        {pageNumbers.map((number) => (
          <button
            key={number}
            onClick={() => onPageChange(number)}
            className={`paginator-page ${
              currentPage === number ? 'active' : ''
            }`}
          >
            {number}
          </button>
        ))}
      </div>

      <button
        className="paginator-button"
        onClick={() => onPageChange(currentPage + 1)}
        disabled={currentPage === totalPages}
      >
        Next
      </button>
    </div>
  );
};
