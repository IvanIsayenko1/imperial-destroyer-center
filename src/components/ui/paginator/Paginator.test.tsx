import { render, screen, fireEvent } from "@testing-library/react";
import { Paginator } from "./Paginator";

describe("Paginator component", () => {
  it("should render page numbers correctly", () => {
    render(
      <Paginator
        currentPage={1}
        totalElements={50}
        elementsPerPage={10}
        onPageChange={() => {}}
      />
    );
    const pageNumbers = screen.getAllByRole("button", { name: /[0-9]/ });
    expect(pageNumbers).toHaveLength(5); // 50 elements, 10 per page, so 5 pages in total
  });

  it("should highlight the current page as active", () => {
    render(
      <Paginator
        currentPage={2}
        totalElements={50}
        elementsPerPage={10}
        onPageChange={() => {}}
      />
    );
    const currentPageButton = screen.getByRole("button", { name: "2" });
    expect(currentPageButton).toHaveClass("active");
  });

  it('should disable "Previous" button on the first page', () => {
    render(
      <Paginator
        currentPage={1}
        totalElements={50}
        elementsPerPage={10}
        onPageChange={() => {}}
      />
    );
    const prevButton = screen.getByText("Previous");
    expect(prevButton).toBeDisabled();
  });

  it('should disable "Next" button on the last page', () => {
    render(
      <Paginator
        currentPage={5}
        totalElements={50}
        elementsPerPage={10}
        onPageChange={() => {}}
      />
    );
    const nextButton = screen.getByText("Next");
    expect(nextButton).toBeDisabled();
  });

  it("should call onPageChange when a page number is clicked", () => {
    const onPageChange = vi.fn();
    render(
      <Paginator
        currentPage={1}
        totalElements={50}
        elementsPerPage={10}
        onPageChange={onPageChange}
      />
    );

    const pageNumberButton = screen.getByRole("button", { name: "3" });
    fireEvent.click(pageNumberButton);
    expect(onPageChange).toHaveBeenCalledWith(3);
  });

  it('should call onPageChange when the "Previous" button is clicked', () => {
    const onPageChange = vi.fn();
    render(
      <Paginator
        currentPage={2}
        totalElements={50}
        elementsPerPage={10}
        onPageChange={onPageChange}
      />
    );

    const prevButton = screen.getByText("Previous");
    fireEvent.click(prevButton);
    expect(onPageChange).toHaveBeenCalledWith(1);
  });

  it('should call onPageChange when the "Next" button is clicked', () => {
    const onPageChange = vi.fn();
    render(
      <Paginator
        currentPage={2}
        totalElements={50}
        elementsPerPage={10}
        onPageChange={onPageChange}
      />
    );

    const nextButton = screen.getByText("Next");
    fireEvent.click(nextButton);
    expect(onPageChange).toHaveBeenCalledWith(3);
  });

  it("should match snapshot", () => {
    const { asFragment } = render(
      <Paginator
        currentPage={1}
        totalElements={50}
        elementsPerPage={10}
        onPageChange={() => {}}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
