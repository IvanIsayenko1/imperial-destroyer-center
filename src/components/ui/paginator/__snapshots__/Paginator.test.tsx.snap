// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Paginator component > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="paginator box"
  >
    <button
      class="paginator-button"
      disabled=""
    >
      Previous
    </button>
    <div
      class="paginator-pages"
    >
      <button
        class="paginator-page active"
      >
        1
      </button>
      <button
        class="paginator-page "
      >
        2
      </button>
      <button
        class="paginator-page "
      >
        3
      </button>
      <button
        class="paginator-page "
      >
        4
      </button>
      <button
        class="paginator-page "
      >
        5
      </button>
    </div>
    <button
      class="paginator-button"
    >
      Next
    </button>
  </div>
</DocumentFragment>
`;
